### PR TITLE
P4C-DPDK: Insert table keys generated by compiler closer to the standard and user metadata

### PR DESCRIPTION
**Pull request by usha1830**
DPDK needs all the table keys to come from single structure and all the table keys should fit within 64 bytes inclusive of all holes in between.

For ex: The below program will fail to load onto DPDK pipeline because the gap between two table keys is more than 64 bytes.

```
struct metadata {
   bit<8> tmp1;
   ...
   ...
   // more fields summing up to >=64 bytes
   ...
   bit <8> tmp2;
}

control{
   table {
      key = {
              tmp1 : exact;
               tmp2 : exact;
         }
    …
   }
   …
}
```
This PR inserts the table keys generated by the compiler right after the standard and user metadata to keep them closer to the metadata fields which can possibly be the table keys in the user program

### CODE REVIEW

File: backends/dpdk/backend.cpp [hunk 1]
```diff
@@ -120,6 +121,7 @@ void DpdkBackend::convert(const IR::ToplevelBlock *tlb) {
         new DpdkAsmOptimization,
         new CollectUsedMetadataField(used_fields),
         new RemoveUnusedMetadataFields(used_fields),
+        new ValidateTableKeys()
```
Comment (usha1830) [inline backends/dpdk/backend.cpp:130]: Added a new pass to emit error if the table keys from metadata struct along with the holes in the metadata structure do not fir in 64 bytes

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 2]
```diff
@@ -175,7 +175,21 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
                     IR::IndexedVector<IR::StructField> usedMetadataFields;
                     for (auto field : st->fields) {
                         if (used_fields.count(field->name.name)) {
-                            usedMetadataFields.push_back(field);
+                            // DPDK implements bool and error types as bit<8>
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.cpp:178]: These changes are made to get the correct offset of metadata fields. 
Earlier, bool and error to bit<8> conversion was happening directly while emitting assembly. Now added this while removing the unused metadata to get the right offset in the next pass.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 3]
```diff
@@ -194,4 +208,63 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
...
+bool ValidateTableKeys::isMetadataStruct(const IR::Type_Struct *st) {
+       for (auto anno : st->annotations->annotations) {
+            if (anno->name == "__metadata__") {
+                return true;
+            }
+       }
+       return false;
+}
+
+bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.cpp:220]: This pass is added in the postgen phase and hence works on the DPDK IR.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 4]
```diff
@@ -194,4 +208,63 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
...
+    const IR::DpdkStructType *metaStruct;
+    for (auto st : p->structType) {
+        if (isMetadataStruct(st)) {
+            metaStruct = st;
```
Comment (psivanup) [inline backends/dpdk/dpdkAsmOpt.cpp:224]: Are there multiple metadata struct possible? Is it necessary to overwrite metaStruct? Can we break after first find?

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 5]
```diff
@@ -194,4 +208,63 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
...
+                        field_type->is<IR::Type_Error>()) {
+                        size_max_field = 8;
+                    } else if (auto t = field_type->to<IR::Type_Name>()) {
+                        if (t->path->name == "error") {
+                            size_max_field = 8;
+                        } else {
+                            BUG("Unexpected type %1%", t->path->name);
+                        }
+                    } else {
+                        BUG("Unexpected type %1%",field_type->node_type_name());
```
Comment (psivanup) [inline backends/dpdk/dpdkAsmOpt.cpp]: space after comma required?

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 6]
```diff
@@ -194,4 +208,63 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
...
+                    }
+                 }
+             }
+            if ((max + size_max_field - min) > 64 * 8) {
+                ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,"%1%: All table keys together "
+                        "with holes in the underlying structure should fit in 64 bytes",tbl->name);
```
Comment (psivanup) [inline backends/dpdk/dpdkAsmOpt.cpp]: space after comma required?

If it is ok to stop at first error, you can break.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 7]
```diff
@@ -194,4 +208,63 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
...
+            }
+        }
+    }
+    return false;
```
Comment (psivanup) [inline backends/dpdk/dpdkAsmOpt.cpp:270]: Is returning false always expected?

**Review by psivanup — CHANGES_REQUESTED** [review]

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 7 above)
Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.cpp:270]: Yes. This is the only visitor and we don't need further traversal. Please correct me if I am wrong, returning false from an Inspector pass stops further traversal in the IR tree.

Comment (psivanup) [reply backends/dpdk/dpdkAsmOpt.cpp:270]: Oh ok.

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 4 above)
Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.cpp:224]: updated.

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 6 above)
Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.cpp]: Updated.

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 5 above)
Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.cpp]: Updated

**Comment by usha1830** [comment]
@psivanup Addressed review comments. Please check if it looks fine

**Review by psivanup — APPROVED** [review]
The changes look fine to me.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 8]
```diff
@@ -194,4 +208,65 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
...
+    for (auto tbl : p->tables) {
+        int min, max, size_max_field = 0;
+        auto keys = tbl->match_keys;
+        if (!keys || keys->keyElements.size() == 0) {
+            return false;
+        }
+        min = max = -1;
+        for (auto key : keys->keyElements) {
+            BUG_CHECK(key->expression->is<IR::Member>(), "Table keys must be a structure field. "
+                                                          "%1% is not a structure field");
```
Comment (github-sajan) [inline backends/dpdk/dpdkAsmOpt.cpp]: Parameter missing for %1% in BUG_CHECK ?

Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.cpp]: @github-sajan Updated. Good catch. Thanks!

**Review by github-sajan — APPROVED** [review]
Looks good to me

**Pull request merged**